src/main/java/org/opensearch/ad/transport/ADResultBulkTransportAction.java

```diff
@@ -94,42 +94,48 @@ protected void doExecute(Task task, ADResultBulkRequest request, ActionListener<
+            for (ResultWriteRequest resultWriteRequest : results) {
+                addResult(bulkRequest, resultWriteRequest.getResult(), resultWriteRequest.getResultIndex());
             }
         } else if (indexingPressurePercent <= hardLimit) {
             // exceed soft limit (60%) but smaller than hard limit (90%)
             float acceptProbability = 1 - indexingPressurePercent;
-            for (AnomalyResult result : results) {
+            for (ResultWriteRequest resultWriteRequest : results) {
+                AnomalyResult result = resultWriteRequest.getResult();
                 if (result.isHighPriority() || random.nextFloat() < acceptProbability) {
-                    addResult(bulkRequest, result);
+                    addResult(bulkRequest, result, resultWriteRequest.getResultIndex());
                 }
             }
         } else {
             // if exceeding hard limit, only index non-zero grade or error result
-            for (AnomalyResult result : results) {
+            for (ResultWriteRequest resultWriteRequest : results) {
+                AnomalyResult result = resultWriteRequest.getResult();
                 if (result.isHighPriority()) {
-                    addResult(bulkRequest, result);
+                    addResult(bulkRequest, result, resultWriteRequest.getResultIndex());
                 }
             }
         }
 
         if (bulkRequest.numberOfActions() > 0) {
-            client.execute(BulkAction.INSTANCE, bulkRequest, ActionListener.<BulkResponse>wrap(bulkResponse -> {
+            client.execute(BulkAction.INSTANCE, bulkRequest, ActionListener.wrap(bulkResponse -> {
                 List<IndexRequest> failedRequests = BulkUtil.getFailedIndexRequest(bulkRequest, bulkResponse);
                 listener.onResponse(new ADResultBulkResponse(failedRequests));
-            }, listener::onFailure));
+            }, e -> {
+                LOG.error("Failed to bulk index AD result", e);
+                listener.onFailure(e);
+            }));
         } else {
             listener.onResponse(new ADResultBulkResponse());
         }
     }
 
-    private void addResult(BulkRequest bulkRequest, AnomalyResult result) {
+    private void addResult(BulkRequest bulkRequest, AnomalyResult result, String resultIndex) {
         try (XContentBuilder builder = jsonBuilder()) {
-            IndexRequest indexRequest = new IndexRequest(indexName).source(result.toXContent(builder, RestHandlerUtils.XCONTENT_WITH_TYPE));
+            String index = resultIndex == null ? indexName : resultIndex;
+            IndexRequest indexRequest = new IndexRequest(index).source(result.toXContent(builder, RestHandlerUtils.XCONTENT_WITH_TYPE));
             bulkRequest.add(indexRequest);
         } catch (IOException e) {
-            LOG.error(String.format(Locale.ROOT, "Failed to prepare bulk %s", indexName), e);
+            LOG.error(String.format(Locale.ROOT, "Failed to prepare bulk %s", resultIndex), e);
```

Comment:
From line 134, seems `resultIndex` could be null, like in the default result index case. Can tune this error statement to be either `indexName` or `resultIndex`

Reply:
good catch, will fix